File: 30.SubstringConcatentionAllWords/SwCoAW.ts
function findSubstring(s: string, words: string[]): number[] {
  if (!s || !words.length) return [];

  const wordLength = words[0].length;
  const wordCount = words.length;
  const totalLength = wordLength * wordCount;
  const wordMap: { [key: string]: number } = {};
  for (const word of words) {
    wordMap[word] = (wordMap[word] || 0) + 1;
  }
  const result: number[] = [];

  for (let offset = 0; offset < wordLength; offset++) {
    let left = offset;
    let right = offset;
    let seen: { [key: string]: number } = {};
    let count = 0;

    while (right + wordLength <= s.length) {
      const word = s.substring(right, right + wordLength);
      right += wordLength;
      if (wordMap[word]) {
        seen[word] = (seen[word] || 0) + 1;
        count++;
        while (seen[word] > wordMap[word]) {
          const leftWord = s.substring(left, left + wordLength);
          seen[leftWord]--;
          left += wordLength;
          count--;
        }
        if (count === wordCount) {
          result.push(left);
          const leftWord = s.substring(left, left + wordLength);
          seen[leftWord]--;
          left += wordLength;
          count--;
        }
      } else {
        seen = {};
        count = 0;
        left = right;
      }
    }
  }
  return result;
}

File: 30.SubstringConcatentionAllWords/SwCoAW.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words:
            return []

        word_length = len(words[0])
        word_count = {}
        for word in words:
            word_count[word] = word_count.get(word, 0) + 1
        num_words = len(words)
        total_length = word_length * num_words
        n = len(s)
        result = []

        for offset in range(word_length):
            left = offset
            right = offset
            curr_count = {}
            count = 0
            while right + word_length <= n:
                word = s[right:right + word_length]
                right += word_length
                if word in word_count:
                    curr_count[word] = curr_count.get(word, 0) + 1
                    count += 1
                    while curr_count[word] > word_count[word]:
                        left_word = s[left:left + word_length]
                        curr_count[left_word] -= 1
                        left += word_length
                        count -= 1
                    if count == num_words:
                        result.append(left)
                else:
                    curr_count.clear()
                    count = 0
                    left = right
        return result